### ISSUE TITLE
Add support for xarray?

### ISSUE THREAD

**Issue by MaximilianR** [Paradigm4/SciDB-Py#90]
I just came across SciDB and it looks very impressive.

We use xarray (https://github.com/pydata/xarray) extensively, whose objects seem to share a lot of characteristics with SciDB.

A high level idea: Could SciDB be used as a backend to xarray? This would be much nicer to work with than numpy arrays. And some of the abstraction is already built into xarray, as out-of-memory dask arrays are a possible backend.

CC @shoyer

**Comment by shoyer** [comment]
We've figured out how to plug-in an alternative numpy-like array system (dask array) to xarray, but we don't have a way to plug-in alternate backends for index operations (like alignment and grouping) as an alternative to `pandas.Index`. That would be necessary to take full advantage of SciDB-Py.

**Comment by jrivers96** [comment]
@shoyer 

I finally got around to looking at xarray this weekend. Very interesting! Are you still looking at being able to plugin indexing operations for arrays? What does your interface look like for this?

**Comment by shoyer** [comment]
> I finally got around to looking at xarray this weekend. Very interesting! Are you still looking at being able to plugin indexing operations for arrays? What does your interface look like for this?

Yes, still open to this, but haven't gotten around to it yet. I think it would be pretty straightforward. For basic indexing functionality (including alignment), would need equivalents for the `get_loc`, `get_indexer` and (optionally) `slice_indexer` methods from `pandas.Index`. The `Index` duck type itself would also need to satisfy some basic array API, including `dtype`/`shape` properties and indexing (`__getitem__`) operations with integers, slices or 1d arrays of integers.
